### PR TITLE
Add support for volume setup via reference volume, add gidfraction parameter

### PR DESCRIPTION
**Pull request by tribal-tec**
Unify the setup of the volume size and resolution in imageSource for
voxelize and livre datasource.

### CODE REVIEW

File: apps/commandLineApplication.h [hunk 1]
```diff
@@ -97,6 +97,7 @@ class CommandLineApplication
               "- BlueConfig: BlueConfig file path\n"
               "              (default: BBPTestData)\n"
               "- target: name of the BlueConfig target (default: CircuitTarget)\n"
+              "- gidfraction: take random cells from a fraction [0,1] of the given target (default: 1)\n"
```
Comment (jafyvilla) [inline apps/commandLineApplication.h]: for consistency: gidFraction or GIDfraction?

File: apps/voxelize/voxelize.cpp [hunk 2]
```diff
@@ -129,30 +128,27 @@ class Voxelize : public CommandLineApplication
...
-        ImageSourcePtr source = params.newImageSource< float >();
-        ::fivox::EventSourcePtr loader = source->getFunctor()->getSource();
-        const fivox::AABBf& bbox = loader->getBoundingBox();
-
-        const fivox::Vector3f& extent( bbox.getSize() +
-                                       params.getExtendDistance() * 2.f );
+        ::fivox::URI uri = getURI();
 
-        size_t size;
+        // for compability
```
Comment (jafyvilla) [inline apps/voxelize/voxelize.cpp]: typo

File: apps/commandLineApplication.h [hunk 3]
```diff
@@ -117,6 +117,10 @@ class CommandLineApplication
               "          of the data events). Changing this parameter will result\n"
               "          in more volumetric data, and therefore more computation\n"
               "          time\n"
+              "- reference: path to a reference volume to take its size and resolution,\n"
+              "             overwrites the 'size' and 'resolution' parameter\n"
+              "- size: size in voxels along the largest dimension of the volume, overwrites the 'resolution' parameter\n"
```
Comment (jafyvilla) [inline apps/commandLineApplication.h]: if already here, shouldn't be removed from apps/voxelize/voxelize.cpp?

File: apps/voxelize/voxelize.cpp [hunk 4]
```diff
@@ -91,8 +91,7 @@ class Voxelize : public CommandLineApplication
               "Type of the data in the output volume "
               "[float (default), int, short, char]" )
             ( "size,s", po::value< size_t >(),
-              "Size of the output volume. If specified, this parameter will "
-              "overwrite the resolution setting in the uri." )
+              "Deprecated; use size in volume URI instead." )
```
Comment (jafyvilla) [inline apps/voxelize/voxelize.cpp]: Ah! here, ignore the other comment :)

File: fivox/imageSource.h [hunk 5]
```diff
@@ -75,6 +75,21 @@ class ImageSource : public itk::ImageSource< TImage >
     /** Enable display of progress bar during voxelization. */
     void showProgress();
 
+    /** Setup size and resolultion of output volume depending on user input. */
+    void setup( const URIHandler& params );
+
+    /** @return the bounding box in of the data in micrometers. */
```
Comment (jafyvilla) [inline fivox/imageSource.h]: remove 'in'

File: doc/Changelog.md [hunk 6]
```diff
@@ -3,12 +3,16 @@ Changelog {#changelog}
 
 # git master {#master}
 
+* [#53](https://github.com/BlueBrain/Fivox/pull/53)
+  Add support for volume setup via reference volume with 'reference' parameter
+  in volume URI
+  Add 'gidfraction' parameter in volume URI
```
Comment (jafyvilla) [inline doc/Changelog.md]: Maybe also add (here and/or in the commit message) that the default resolution is now 1 instead of 10

**Comment by jafyvilla** [comment]
Other than the minor comments above, +1

File: apps/voxelize/voxelize.cpp [hunk 7]
```diff
@@ -129,30 +128,27 @@ class Voxelize : public CommandLineApplication
...
-        size_t size;
+        // for compatibility
         if( _vm.count( "size" ))
-            size = _vm["size"].as< size_t >();
-        else
-        {
-            const fivox::Vector3f& sizeVoxels = extent * params.getResolution();
-            size = (size_t)std::ceil( sizeVoxels.find_max( ));
-        }
+            uri.addQuery( "size", std::to_string( _vm["size"].as< size_t >( )));
```
Comment (eile) [inline apps/voxelize/voxelize.cpp]: Just ```_vm["size"].as< std::string >()```?

File: fivox/imageSource.hxx [hunk 8]
```diff
@@ -78,6 +80,78 @@ void ImageSource< TImage >::PrintSelf(std::ostream & os, itk::Indent indent )
 }
 
 template< typename TImage >
+void ImageSource< TImage >::setup( const URIHandler& params )
+{
+    const std::string& refVolume = params.getReferenceVolume();
+    if( !refVolume.empty( ))
```
Comment (eile) [inline fivox/imageSource.hxx]: swap if/else?

File: fivox/uriHandler.cpp [hunk 9]
```diff
@@ -209,16 +213,29 @@ class URIHandler::Impl
 
     std::string getDyeCurve() const { return _get( "dyecurve" ); }
 
-    float getResolution() const { return _get( "resolution", _resolution ); }
+    float getResolution() const
+    {
+        return _get( "resolution", getFunctorType() == FUNCTOR_DENSITY
+                                   ? 0.0625f : _resolution );
```
Comment (eile) [inline fivox/uriHandler.cpp]: magic? Still needed?

File: apps/voxelize/voxelize.cpp (hunk 7 above)
Comment (tribal-tec) [inline apps/voxelize/voxelize.cpp]: Same train of though as I had, but can't be lexical_casted from the commandline if not specified as string... :) Hence, falling back was working.

File: fivox/uriHandler.cpp (hunk 9 above)
Comment (tribal-tec) [inline fivox/uriHandler.cpp]: Not really magic, it's the 16um/voxel default that makes sense for densities. Same magics as we have for the input ranges of the different sources.

File: doc/Changelog.md (hunk 6 above)
Comment (tribal-tec) [inline doc/Changelog.md]: Done

**Comment by tribal-tec** [comment]
Good now?

**Comment by jafyvilla** [comment]
Good for me

**Pull request merged**